### PR TITLE
JAMES-2013 Extract Tika service from James

### PR DESCRIPTION
**Pull request by aduprat**
The code modification is finished for this task, so I want to share it to you.

But I will handle the remaining things:
- documentation
- take care of docker-compose
- deploy & document Tika on our production

### CODE REVIEW

**Comment by chibenwa** [comment]
I don't like the idea of running a docker container for thousands of tests...

It is slow (at least on my development machine)... Why can't we have :

 - An EmbeddedTikaTextExtractor
 - We write basic tests for both
 - We use the distant TikaTextExtractor in some well identified use cases (to ensure it works) (maybe ~10 tests)
 - We can then use embedded tika everywhere else to speed up testing...

What do you think about this?

File: mailbox/tika/src/main/java/org/apache/james/mailbox/tika/TikaHttpClientImpl.java [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+        return new URIBuilder()
+                .setHost(tikaConfiguration.getHost())
+                .setPort(tikaConfiguration.getPort())
+                .setScheme("http")
+                .build();
+    }
+
+    @Override
+    public String metadataAndContent(InputStream inputStream, String contentType) throws TikaException {
+        return executePut(uri.resolve(RMETA_AS_TEXT_ENDPOINT), inputStream, contentType);
```
Comment (chibenwa) [inline mailbox/tika/src/main/java/org/apache/james/mailbox/tika/TikaHttpClientImpl.java]: Do we need to do the uri.resolve on each request?

Comment (chibenwa) [inline mailbox/tika/src/main/java/org/apache/james/mailbox/tika/TikaHttpClientImpl.java]: Then you can easily inline executePut

File: mailbox/tika/src/main/java/org/apache/james/mailbox/tika/TikaTextExtractor.java [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+    
+    private static class ContentAndMetadata {
+
+        private static final String TIKA_HEADER = "X-TIKA";
+        private static final String CONTENT_METADATA_HEADER_NAME = TIKA_HEADER + ":content";
+
+        public static ContentAndMetadata from(Map<String, List<String>> contentAndMetadataMap) {
+            String onlySpaces = null;
+            return new ContentAndMetadata(content(contentAndMetadataMap, onlySpaces),
+                    Maps.filterEntries(contentAndMetadataMap, new Predicate<Map.Entry<String, List<String>>>() {
```
Comment (quynhn) [inline mailbox/tika/src/main/java/org/apache/james/mailbox/tika/TikaTextExtractor.java]: It should be easier for reading if extract Predicate to constant or function I think

File: mailbox/tika/src/test/java/org/apache/james/mailbox/tika/TikaContainer.java [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+import com.jayway.awaitility.core.ConditionFactory;
+
+public class TikaContainer extends ExternalResource {
+    
+    private static final int DEFAULT_TIKA_PORT = 9998;
+    private static final int DEFAULT_TIMEOUT_IN_MS = Ints.checkedCast(TimeUnit.MINUTES.toMillis(3));
+
+    private static final ConditionFactory CALMLY_AWAIT = Awaitility.with()
+            .pollInterval(Duration.FIVE_HUNDRED_MILLISECONDS)
+            .and().with()
```
Comment (quynhn) [inline mailbox/tika/src/test/java/org/apache/james/mailbox/tika/TikaContainer.java]: one more break after and()?

File: server/container/guice/cassandra-guice/src/main/java/org/apache/james/modules/mailbox/TikaMailboxModule.java [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+        try {
+            PropertiesConfiguration configuration = propertiesProvider.getConfiguration(TIKA_CONFIGURATION_NAME);
+            return TikaConfiguration.builder()
+                    .host(configuration.getString(TIKA_HOST, DEFAULT_HOST))
+                    .port(configuration.getInt(TIKA_PORT, DEFAULT_PORT))
+                    .timeoutInMillis(configuration.getInt(TIKA_TIMEOUT_IN_MS, DEFAULT_TIMEOUT_IN_MS))
+                    .build();
+        } catch (FileNotFoundException e) {
+            LOGGER.warn("Could not find " + TIKA_CONFIGURATION_NAME + " configuration file. Using 127.0.0.1:9998 as contact point");
+            return TikaConfiguration.builder()
```
Comment (chibenwa) [inline server/container/guice/cassandra-guice/src/main/java/org/apache/james/modules/mailbox/TikaMailboxModule.java]: If no config, maybe we can fallback to an EmbeddedTikaTextExtractor, with appropriate warnings positioned.

This would provide me, who run a single James server for a few 10s users on a RasberryPy to have to install yet another Java service.

We can:
 - Have an isEmbedded() Method on the config
 - We default saying it's embedded
 - Then we can fallback when provisionning TextExtractor to the Embedded version

What do you think of this?

File: server/container/guice/cassandra-guice/src/main/java/org/apache/james/modules/mailbox/TikaMailboxModule.java [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+    private TikaConfiguration getTikaConfiguration(PropertiesProvider propertiesProvider) throws ConfigurationException {
+        try {
+            PropertiesConfiguration configuration = propertiesProvider.getConfiguration(TIKA_CONFIGURATION_NAME);
+            return TikaConfiguration.builder()
+                    .host(configuration.getString(TIKA_HOST, DEFAULT_HOST))
+                    .port(configuration.getInt(TIKA_PORT, DEFAULT_PORT))
+                    .timeoutInMillis(configuration.getInt(TIKA_TIMEOUT_IN_MS, DEFAULT_TIMEOUT_IN_MS))
+                    .build();
+        } catch (FileNotFoundException e) {
+            LOGGER.warn("Could not find " + TIKA_CONFIGURATION_NAME + " configuration file. Using 127.0.0.1:9998 as contact point");
```
Comment (quynhn) [inline server/container/guice/cassandra-guice/src/main/java/org/apache/james/modules/mailbox/TikaMailboxModule.java]: 127.0.0.1:9998 = DEFAULT_HOST + DEFAULT_PORT?

File: mailbox/tika/src/test/java/org/apache/james/mailbox/tika/TikaContainer.java [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+        return DEFAULT_TIKA_PORT;
+    }
+
+    public int getTimeoutInMillis() {
+        return DEFAULT_TIMEOUT_IN_MS;
+    }
+
+    public void start() throws Exception {
+        tika.start();
+        awaitForTika();
```
Comment (mbaechler) [inline mailbox/tika/src/test/java/org/apache/james/mailbox/tika/TikaContainer.java]: there is APIs in testcontainers to do that

File: mailbox/tika/src/test/java/org/apache/james/mailbox/tika/TikaTextExtractorTest.java [hunk 7]
```diff
@@ -89,15 +115,15 @@ public void pdfTest() throws Exception {
         InputStream inputStream = ClassLoader.getSystemResourceAsStream("documents/PDF.pdf");
         assertThat(inputStream).isNotNull();
         assertThat(textExtractor.extractContent(inputStream, "application/pdf", "PDF.pdf").getTextualContent())
-            .isEqualTo("\nThis is an awesome document on libroffice writter !\n\n\n");
+            .isEqualTo("This is an awesome document on libroffice writter !\n\n\n");
```
Comment (mbaechler) [inline mailbox/tika/src/test/java/org/apache/james/mailbox/tika/TikaTextExtractorTest.java]: it's where equality doesn't make sense, I'm sure there are other matchers that would better fit

File: mailbox/tika/src/test/java/org/apache/james/mailbox/tika/TikaTextExtractorTest.java [hunk 8]
```diff
@@ -121,5 +147,38 @@ public void excelTest() throws Exception {
                 "\n" +
                 "\n");
     }
-    
+
+    @Test
+    public void deserializerShouldThrowWhenMoreThanOneNode() throws Exception {
```
Comment (mbaechler) [inline mailbox/tika/src/test/java/org/apache/james/mailbox/tika/TikaTextExtractorTest.java]: given the deserialization code complexity, you need much more tests.
Maybe doing unit tests on the deserializer would be a better idea

File: server/container/guice/cassandra-guice/src/main/java/org/apache/james/modules/mailbox/TikaMailboxModule.java (hunk 5 above)
Comment (mbaechler) [inline server/container/guice/cassandra-guice/src/main/java/org/apache/james/modules/mailbox/TikaMailboxModule.java]: use logger placeholder '{}'

File: server/container/guice/cassandra-guice/src/main/java/org/apache/james/modules/mailbox/TikaMailboxModule.java (hunk 4 above)
Comment (mbaechler) [inline server/container/guice/cassandra-guice/src/main/java/org/apache/james/modules/mailbox/TikaMailboxModule.java]: It's probably not a good idea to embedded tika for everybody. It pulls lot of code in-process and will not make everybody happy.
Document conversion is not that important, we can leave without if not configured, that's better than poor performance.

File: server/protocols/jmap/src/test/java/org/apache/james/jmap/model/MessageFactoryTest.java [hunk 9]
```diff
@@ -58,16 +63,31 @@
...
-    public void setUp() {
-        htmlTextExtractor = new MailboxBasedHtmlTextExtractor(new TikaTextExtractor());
+    public void setUp() throws Exception {
+        tika.start();
+        TikaTextExtractor textExtractor = new TikaTextExtractor(new TikaHttpClientImpl(TikaConfiguration.builder()
+                .host(tika.getIp())
+                .port(tika.getPort())
+                .timeoutInMillis(tika.getTimeoutInMillis())
+                .build()));
+        htmlTextExtractor = new MailboxBasedHtmlTextExtractor(textExtractor);
```
Comment (mbaechler) [inline server/protocols/jmap/src/test/java/org/apache/james/jmap/model/MessageFactoryTest.java]: we should not use tika for html extraction, we have jsoup converter for that AFAIK

**Comment by mbaechler** [comment]
To answer benoit : I think running tika in a container can actually be faster than in-process : 
1. it's stateless so we can use the same container for many tests
2. it will speedup james by removing bunch of dependencies
Let's get timing before going ahead with this debate, don't you think ?

File: mailbox/store/src/test/resources/eml/nonTextual.json [hunk 10]
```diff
@@ -133,7 +133,6 @@
         "meta:table-count":["0"],
         "Creation-Date":["2015-06-18T12:41:25.197399866"],
         "xmpTPg:NPages":["1"],
-        "resourceName":["toto.odt"],
```
Comment (aduprat) [inline mailbox/store/src/test/resources/eml/nonTextual.json]: This header is always added by the embedded Tika, not by the /rmeta endpoint.
When using the Tika as a server, this header is given by the /detect endpoint:
https://wiki.apache.org/tika/TikaJAXRS#Detector_Resource
https://github.com/apache/tika/blob/0d8f030f02830a39e004fbfbd6989128ec2047d7/tika-server/src/main/java/org/apache/tika/server/resource/DetectorResource.java#L51

File: mailbox/tika/src/main/java/org/apache/james/mailbox/tika/TikaTextExtractor.java [hunk 11]
```diff
@@ -0,0 +1,170 @@
...
+
+    private static class ContentAndMetadataDeserializer extends JsonDeserializer<ContentAndMetadata> {
+
+        @Override
+        public ContentAndMetadata deserialize(JsonParser jsonParser, DeserializationContext deserializationContext) throws IOException, JsonProcessingException {
+            TreeNode treeNode = jsonParser.getCodec().readTree(jsonParser);
+            Preconditions.checkState(treeNode.isArray() && treeNode.size() == 1, "The response should have only one element");
+            Preconditions.checkState(treeNode.get(0).isObject(), "The element should be a Json object");
+            ObjectNode node = (ObjectNode) treeNode.get(0);
+            return ContentAndMetadata.from(ImmutableList.copyOf(node.fields()).stream()
```
Comment (aduprat) [inline mailbox/tika/src/main/java/org/apache/james/mailbox/tika/TikaTextExtractor.java]: _fields_ returns *Iterator<Map.Entry<String, JsonNode>>*

File: mailbox/tika/src/main/java/org/apache/james/mailbox/tika/TikaTextExtractor.java [hunk 12]
```diff
@@ -0,0 +1,170 @@
...
+                .map(entry -> Pair.of(entry.getKey(), asListOfString(entry.getValue())))
+                .collect(Collectors.toMap(Pair::getLeft, Pair::getRight)));
+        }
+
+        private List<String> asListOfString(JsonNode jsonNode) {
+            if (jsonNode.isArray()) {
+                return ImmutableList.copyOf(jsonNode.elements()).stream()
```
Comment (aduprat) [inline mailbox/tika/src/main/java/org/apache/james/mailbox/tika/TikaTextExtractor.java]: _elements_ returns _Iterator<JsonNode>_

File: mailbox/tika/src/main/java/org/apache/james/mailbox/tika/TikaTextExtractor.java [hunk 13]
```diff
@@ -0,0 +1,170 @@
...
+    }
+    
+    private static class ContentAndMetadata {
+
+        private static final String TIKA_HEADER = "X-TIKA";
+        private static final String CONTENT_METADATA_HEADER_NAME = TIKA_HEADER + ":content";
+
+        public static ContentAndMetadata from(Map<String, List<String>> contentAndMetadataMap) {
+            String onlySpaces = null;
+            return new ContentAndMetadata(content(contentAndMetadataMap, onlySpaces),
```
Comment (aduprat) [inline mailbox/tika/src/main/java/org/apache/james/mailbox/tika/TikaTextExtractor.java]: Comes from *StringUtils.stripStart*
```
public static String stripStart(String str,
                                String stripChars)
Strips any of a set of characters from the start of a String.

A null input String returns null. An empty string ("") input returns the empty string.

If the stripChars String is null, whitespace is stripped as defined by Character.isWhitespace(char).
```

File: mailbox/tika/src/main/java/org/apache/james/mailbox/tika/TikaTextExtractor.java (hunk 11 above)
Comment (mbaechler) [inline mailbox/tika/src/main/java/org/apache/james/mailbox/tika/TikaTextExtractor.java]: ok, copy is really overkill, but let say it's not a problem right now

File: mailbox/tika/src/main/java/org/apache/james/mailbox/tika/TikaTextExtractor.java (hunk 13 above)
Comment (mbaechler) [inline mailbox/tika/src/main/java/org/apache/james/mailbox/tika/TikaTextExtractor.java]: ok but there, you are passing in to you own function

File: mailbox/tika/src/main/java/org/apache/james/mailbox/tika/TikaTextExtractor.java [hunk 14]
```diff
@@ -0,0 +1,170 @@
...
+                    Maps.filterEntries(contentAndMetadataMap, new Predicate<Map.Entry<String, List<String>>>() {
+
+                        @Override
+                        public boolean apply(Map.Entry<String, List<String>> entry) {
+                            return !entry.getKey().startsWith(TIKA_HEADER);
```
Comment (aduprat) [inline mailbox/tika/src/main/java/org/apache/james/mailbox/tika/TikaTextExtractor.java]: Because the Tikas server endpoint */rmeta* is adding some Tika headers (process duration, ...).
I don't think we want to index them .

File: mailbox/tika/src/main/java/org/apache/james/mailbox/tika/TikaTextExtractor.java [hunk 15]
```diff
@@ -0,0 +1,170 @@
...
+                return null;
+            }
+            return StringUtils.stripStart(content.get(0), onlySpaces);
+        }
+
+        private final String content;
+        private final Map<String, List<String>> metadata;
+
+        private ContentAndMetadata(String content, Map<String, List<String>> metadata) {
+            this.content = content;
```
Comment (aduprat) [inline mailbox/tika/src/main/java/org/apache/james/mailbox/tika/TikaTextExtractor.java]: Yes, that's the behaviour of the previous Tika release; I don't want to change that as it will lead to lot of modification (JMAP, ...)

File: mailbox/tika/src/test/java/org/apache/james/mailbox/tika/TikaContainer.java (hunk 6 above)
Comment (aduprat) [inline mailbox/tika/src/test/java/org/apache/james/mailbox/tika/TikaContainer.java]: Thx, I don't know that.

File: mailbox/tika/src/main/java/org/apache/james/mailbox/tika/TikaTextExtractor.java (hunk 11 above)
Comment (aduprat) [inline mailbox/tika/src/main/java/org/apache/james/mailbox/tika/TikaTextExtractor.java]: I've corrected this but not pushed yet.

File: mailbox/tika/src/main/java/org/apache/james/mailbox/tika/TikaTextExtractor.java (hunk 14 above)
Comment (mbaechler) [inline mailbox/tika/src/main/java/org/apache/james/mailbox/tika/TikaTextExtractor.java]: the method name should explain the "why"

File: mailbox/tika/src/test/java/org/apache/james/mailbox/tika/TikaContainer.java (hunk 6 above)
Comment (aduprat) [inline mailbox/tika/src/test/java/org/apache/james/mailbox/tika/TikaContainer.java]: but leads to unstable tests...

File: server/protocols/jmap/src/test/java/org/apache/james/jmap/model/MessageFactoryTest.java (hunk 9 above)
Comment (aduprat) [inline server/protocols/jmap/src/test/java/org/apache/james/jmap/model/MessageFactoryTest.java]: The goal of this task is to extract, not change all the tests behavior.
I'm afraid of too much responsibility on this issue, then it will take a long time to be approved/merged.

File: mailbox/tika/src/test/java/org/apache/james/mailbox/tika/TikaContainer.java (hunk 6 above)
Comment (mbaechler) [inline mailbox/tika/src/test/java/org/apache/james/mailbox/tika/TikaContainer.java]: I don't get it, it does exactly the same, are you sure we are talking about the same thing ? https://www.testcontainers.org/usage/options.html#startup-timeout see WaitFor (that does http request in a loop waiting for the right http status code)

File: mailbox/tika/src/main/java/org/apache/james/mailbox/tika/TikaTextExtractor.java (hunk 15 above)
Comment (mbaechler) [inline mailbox/tika/src/main/java/org/apache/james/mailbox/tika/TikaTextExtractor.java]: ok

File: server/protocols/jmap/src/test/java/org/apache/james/jmap/model/MessageFactoryTest.java (hunk 9 above)
Comment (mbaechler) [inline server/protocols/jmap/src/test/java/org/apache/james/jmap/model/MessageFactoryTest.java]: ok, you are right, but nonetheless, I just discovered a source of useless waste, I'll create a ticket for that.

**Comment by aduprat** [comment]
test this please

**Comment by aduprat** [comment]
test this please

File: mailbox/tika/src/test/java/org/apache/james/mailbox/tika/TikaContainer.java (hunk 6 above)
Comment (aduprat) [inline mailbox/tika/src/test/java/org/apache/james/mailbox/tika/TikaContainer.java]: Was not so easy ;)

**Comment by aduprat** [comment]
test this please

**Comment by chibenwa** [comment]
If this container is stateless, then :

 - we can point to a specific IP of this always launched container
 - if absent start the docker

This would allow CI to never have to start tika containers. What do think ?

File: server/container/guice/cassandra-guice/src/test/java/org/apache/james/GuiceTikaRule.java [hunk 16]
```diff
@@ -0,0 +1,53 @@
...
+import org.apache.james.modules.TestTikaModule;
+import org.junit.runner.Description;
+import org.junit.runners.model.Statement;
+
+import com.google.inject.Module;
+
+
+public class GuiceTikaRule implements GuiceModuleTestRule {
+
+    private TikaContainer tika;
```
Comment (chibenwa) [inline server/container/guice/cassandra-guice/src/test/java/org/apache/james/GuiceTikaRule.java]: Supplier + memoize from GUAVA

File: server/container/guice/cassandra-guice/src/test/java/org/apache/james/GuiceTikaRule.java [hunk 17]
```diff
@@ -0,0 +1,53 @@
...
+
+    @Override
+    public Statement apply(Statement base, Description description) {
+        return new Statement() {
```
Comment (chibenwa) [inline server/container/guice/cassandra-guice/src/test/java/org/apache/james/GuiceTikaRule.java]: Here : 

```
return tikaSupplier.get().apply(base, description);
```

Tika is started but not chained, thus JUNIT could not stop the container, explaining your 20 tika server started on CI, IMO

Anyway, it's even cleaner from a code perspective, so ....

File: server/container/guice/cassandra-guice/src/test/java/org/apache/james/GuiceTikaRule.java [hunk 18]
```diff
@@ -0,0 +1,53 @@
...
+        };
+    }
+
+    @Override
+    public void await() {
+    }
+
+    @Override
+    public Module getModule() {
+        return new TestTikaModule(tika);
```
Comment (chibenwa) [inline server/container/guice/cassandra-guice/src/test/java/org/apache/james/GuiceTikaRule.java]: With the supplier I would avoid creating a module with a null reference ;-)

**Comment by mbaechler** [comment]
Let's get timing before going ahead with this debate, don't you think ?

File: mailbox/elasticsearch/src/test/java/org/apache/james/mailbox/elasticsearch/json/MessageToElasticSearchJsonTest.java [hunk 19]
```diff
@@ -63,11 +69,22 @@
...
     private Date date;
     private PropertyBuilder propertyBuilder;
 
+    @Rule
+    public TikaContainer tika = new TikaContainer();
+
     @Before
     public void setUp() throws Exception {
+        tika.start();
+        textExtractor = new TikaTextExtractor(new TikaHttpClientImpl(TikaConfiguration.builder()
```
Comment (strokyl) [inline mailbox/elasticsearch/src/test/java/org/apache/james/mailbox/elasticsearch/json/MessageToElasticSearchJsonTest.java]: Could be great to put that in a rule too so we will just need to do `tikaRule.getTextExtractor`

File: mailbox/tika/src/main/java/org/apache/james/mailbox/tika/TikaTextExtractor.java (hunk 12 above)
Comment (strokyl) [inline mailbox/tika/src/main/java/org/apache/james/mailbox/tika/TikaTextExtractor.java]: But if you collecting you already converting now ? By the way why aren't you directly collecting to immutable list ?

File: mailbox/tika/src/test/java/org/apache/james/mailbox/tika/TikaContainer.java [hunk 20]
```diff
@@ -0,0 +1,94 @@
...
+            URI uri = new URIBuilder()
+                    .setHost(tika.getIp())
+                    .setPort(DEFAULT_TIKA_PORT)
+                    .setScheme("http")
+                    .build();
+            return Executor.newInstance()
+                .execute(Request.Get(uri.resolve("/tika")))
+                .returnResponse()
+                .getStatusLine()
+                .getStatusCode() == 200;
```
Comment (strokyl) [inline mailbox/tika/src/test/java/org/apache/james/mailbox/tika/TikaContainer.java]: I would check that is in [200..299]

File: server/protocols/jmap/pom.xml [hunk 21]
```diff
@@ -275,6 +287,11 @@
                     <scope>test</scope>
                 </dependency>
                 <dependency>
+                    <groupId>com.jayway.awaitility</groupId>
+                    <artifactId>awaitility</artifactId>
```
Comment (strokyl) [inline server/protocols/jmap/pom.xml]: Before in the PR you introduce this dependence and specify a version for it, you did not it here ? Why did you need to specify before and not here ?

File: server/protocols/jmap/src/test/java/org/apache/james/jmap/methods/GetMessagesMethodTest.java [hunk 22]
```diff
@@ -116,7 +124,13 @@ public boolean isSameUser(String username) {
     @Before
     public void setup() throws Exception {
         clientId = ClientId.of("#0");
-        HtmlTextExtractor htmlTextExtractor = new MailboxBasedHtmlTextExtractor(new TikaTextExtractor());
+        tika.start();
+        TikaTextExtractor textExtractor = new TikaTextExtractor(new TikaHttpClientImpl(TikaConfiguration.builder()
```
Comment (strokyl) [inline server/protocols/jmap/src/test/java/org/apache/james/jmap/methods/GetMessagesMethodTest.java]: Could be great to move that in a rule so we do not need to do it over and over.

**Comment by chibenwa** [comment]
See https://github.com/linagora/james-project/issues/811
